feat(i18n): add documentation for custom paths (#5526)
withastro/docs: src/content/docs/en/guides/internationalization.mdx

Depending on your deploy host, you may discover transformations in URL paths, so check your deployed site to determine the best syntax for your `locale` values.

### Custom locale paths

In addition to defining your site's supported `locales` as strings (e.g. "en", "pt-br"), Astro also allows you to map an arbitrary number of [browser-recognized language `codes`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#syntax) to a custom URL `path`. While locales can be strings of any format as long as they correspond to your project folder structure, `codes` must follow the browser's accepted syntax.

Pass an object to the `locales` array with a `path` key to define a custom URL prefix, and `codes` to indicate the languages mapped to this URL. In this case, your `/[locale]/` folder name must match exactly the value of the `path` and your URLs will be generated using the `path` value.

> This is useful if you support multiple variations of language (e.g. `"fr"`, `"fr-BR"`, and `"fr-CA"`) and you want to have all these locales mapped under the same URL `/fr/`, or even customize it entirely (e.g. `/french/`):

Reviewer: ```suggestion
This is useful if you support multiple variations of a language (e.g. `"fr"`, `"fr-BR"`, and `"fr-CA"`) and you want to have all these locales mapped under the same URL `/fr/`, or even customize it entirely (e.g. `/french/`):
```

`getAbsoluteLocaleUrlList(locale: string, options?: GetLocaleOptions): string[]`

Use this like [`getAbsoluteLocaleUrl`](#getabsolutelocaleurl) to return a list of absolute paths for all the locales.

### `getPathByLocale(locale: string): string`

A function that return the `path` associated to a locale (defined as code). It's particularly useful in case you use [refined locales](#refined-locales).

> ### Refined locales

Reviewer: ```suggestion
### Custom locale paths
```
Initial thinking! We can see what we like here.

> Astro allows to map an arbitrary number of locales to a particular URL path. The `locales` array can accept an object, where the `path` key is what you'll see in the URL, and `codes` represents the languages mapped to this URL.

Author: Updated the suggestion to fix a typo

```js title="astro.config.mjs" del={6} ins={7-10}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      locales: ["es", "en", {
          path: "french", // not slash needed
          codes: ["fr", "fr-BR", "fr-CA"]
      }],
      routingStrategy: "prefix-always"
    }
  }
})
```

If you rely on the [`astro:i18n` virtual module](#virtual-module-astroi18n), you'll have to use the `path` to compose the correct URL, for example:

```astro title="src/pages/index.astro"
---
import { getPathByLocale, getRelativeLocaleUrl, getLocaleByPath } from "astro:i18n";

getRelativeLocaleUrl("french", "blog"); // returns /french/blog

// These functions are not particularly useful 
getPathByLocale("fr"); // returns "french"
getPathByLocale("fr-CA"); // returns "french"
getPathByLocale("es"); // returns "es"
getLocaleByPath("french"); // returns "fr" because that's the first code configured
getLocaleByPath("es"); // returns "es"
---
```

> If you rely on [the browser language detection](#browser-language-detection), Astro will use the `codes` to determine the `Astro.preferredLocale` and `Astro.preferredLocaleList`. If so, it's advised to write the codes using the same pattern [used by the browser](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#syntax).

Reviewer: ```suggestion
Astro will use your configured `codes` and [the browser's language detection](#browser-language-detection) to determine `Astro.preferredLocale` and `Astro.preferredLocaleList`. In order to successfully match your visitors' preferences, provide your `codes` using the same pattern [used by the browser](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#syntax).
```

`getAbsoluteLocaleUrlList(locale: string, options?: GetLocaleOptions): string[]`

Use this like [`getAbsoluteLocaleUrl`](#getabsolutelocaleurl) to return a list of absolute paths for all the locales.

### `getPathByLocale(locale: string): string`

A function that returns the `path` associated to one or more `codes` when [custom locale paths](#custom-locale-paths) are configured.

> ### Generatng URLs from custom paths

Reviewer: ```suggestion
### Generating URLs from custom paths
```

> When a `locale` is a custom path, use the `path` as the value for `locale`.

Reviewer: ``````suggestion
When a `locale` is a custom path, use the `path` as the value for `locale`.

For example, use `french` in the i18n helper functions for the `locales` configuration below:

```js
    locales: ["es", "en", {
        path: "french",
        codes: ["fr", "fr-BR", "fr-CA"]
    }],
```
``````
